### PR TITLE
Audio streaming training with masking

### PR DESCRIPTION
**Pull request by saeeddhqan**
By utilizing a mask, we can simulate audio streaming.

### CODE REVIEW

File: ultravox/model/ultravox_model.py [hunk 1]
```diff
@@ -55,6 +55,24 @@ def __init__(self, config: UltravoxConfig):
...
         )
 
+        audio_latency_bsize = config.audio_latency_bsize
+        audio_streaming_mask = None
+        if audio_latency_bsize != -1:
+            audio_latency_nblocks = (self.audio_tower.config.max_source_positions
+                * self.audio_tower.conv1.stride[0]
+                * self.audio_tower.conv2.stride[0]
+            )
+            audio_latency_nblocks //=  2 * audio_latency_bsize
```
[the remark this answers is not in the document]
Comment (farzadab) [reply ultravox/model/ultravox_model.py:65]: Actually, if `max_source_positions`  is not divisible by `audio_latency_bsize`, we might be in trouble, so I'd rather put an `assert` here.

**Review by farzadab — COMMENTED** [review]
Thanks Saeed. I left some comments.

It would be ideal to add some tests, for example to make sure the structure of the mask matches what we expect. If you can think of another test feel free to add one.
I know the repo doesn't have many tests, but we'd like to change that.

**Comment by saeeddhqan** [comment]
@farzadab Hi,
I created a test file for ultravox_model, moved the function, and changed var names. 
I look forward to further improvements.

File: ultravox/model/ultravox_model.py [hunk 2]
```diff
@@ -286,13 +287,16 @@ def _create_audio_tower(
                 audio_tower = ModifiedWhisperEncoder.from_pretrained(
                     config.audio_model_id, torch_dtype=config.torch_dtype
                 )
+                audio_tower.init_latency_mask(
+                    config.audio_latency_block_size, dtype=config.torch_dtype)
```
Comment (farzadab) [inline ultravox/model/ultravox_model.py:291]: Thanks. It would be good to have an assert in the other case as well if block_size is non-zero rather than silently refusing to use it.

File: ultravox/model/ultravox_model.py [hunk 3]
```diff
@@ -529,6 +533,35 @@ class ModifiedWhisperEncoder(
     base_model_prefix = "model.encoder"
     _no_split_modules = ["WhisperEncoderLayer"]
 
+    def init_latency_mask(self, audio_latency_block_size: int, dtype: torch.float):
+        if audio_latency_block_size is None:
+            self.audio_streaming_mask = None
+            return
+        audio_streaming_mask = None
+        # total sequence length
```
Comment (farzadab) [inline ultravox/model/ultravox_model.py:541]:
```suggestion
        # maximum sequence length
```

File: ultravox/model/ultravox_model.py [hunk 4]
```diff
@@ -529,6 +533,35 @@ class ModifiedWhisperEncoder(
...
+
+        audio_latency_nblocks //=  2 * audio_latency_block_size
+        audio_streaming_mask = torch.tril(
+            torch.ones(
+                audio_latency_nblocks, audio_latency_nblocks),
+                diagonal=0,
+            ).repeat_interleave(audio_latency_block_size, dim=0).repeat_interleave(audio_latency_block_size, dim=1)
+        audio_streaming_mask = (1.0 - audio_streaming_mask) * torch.finfo(dtype).min
+        # audio_streaming_mask[audio_streaming_mask == 0] = float('-inf')
+        # audio_streaming_mask[audio_streaming_mask == 1] = 0
```
Comment (farzadab) [inline ultravox/model/ultravox_model.py:561]: nit: please remove commented code

File: ultravox/training/configs/streaming_tinyllama.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# SLM with ultravox & llama3.1, trained wtih knowledge distillation.
```
Comment (farzadab) [inline ultravox/training/configs/streaming_tinyllama.yaml:1]: Do you have any experimental results with this in Ultravox repo?
Also the experiment name mentions "1s" whereas the `block_size` is `null` which seems incorrect.

File: ultravox/model/ultravox_model_test.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import pytest
+import torch
+from transformers import WhisperConfig
+from transformers.models.whisper import modeling_whisper as whisper
+from ultravox.model import ultravox_model
+
+class TestModifiedWhisperEncoder:
```
Comment (farzadab) [inline ultravox/model/ultravox_model_test.py:7]: nit: We don't use classes for our tests. It's probably best to use stand-alone functions for consistency since I don't see the `self` being useful anywhere.

File: ultravox/model/ultravox_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (farzadab) [reply ultravox/model/ultravox_model.py]: I don't think this was answered. The `*2` is still there and I don't see an explanation here or in the code about it. It would be ideal to avoid magic numbers as much as possible.

**Review by farzadab — COMMENTED** [review]
Thanks. I left a few nit comments.
Please make sure to run `just format` to run formatting and use `just test` to make sure the tests pass.

File: ultravox/training/configs/streaming_tinyllama.yaml (hunk 5 above)
Comment (saeeddhqan) [reply ultravox/training/configs/streaming_tinyllama.yaml]: Yes, I did all the experiments with the forked repo; and I normally run a training with a few steps after conducted changes to make sure things work smoothly.

**Comment by saeeddhqan** [comment]
@farzadab, I utilized 'just test', and 'just format', and corrected mentioned issues.

**Review by farzadab — APPROVED** [review]
Assuming the tests pass it LGTM. Thanks Saeed!

File: ultravox/training/config_base.py [hunk 7]
```diff
@@ -106,6 +106,9 @@ def get_val_sets(self) -> List[DatasetOptions]:
     # loss function to use
     loss_config: Optional[ultravox_config.LossConfig] = None
 
+    # To simulate audio streaming with masking. None for non-causal, 100 for 1s, 200 for 2s, and so on.
+    audio_latency_block_size: int = None
```
Comment (farzadab) [inline ultravox/training/config_base.py:110]:
```suggestion
    audio_latency_block_size: Optional[int] = None
```

File: ultravox/model/ultravox_model.py [hunk 8]
```diff
@@ -529,6 +544,39 @@ class ModifiedWhisperEncoder(
     base_model_prefix = "model.encoder"
     _no_split_modules = ["WhisperEncoderLayer"]
 
+    def init_latency_mask(self, audio_latency_block_size: int, dtype: torch.float):
```
Comment (farzadab) [inline ultravox/model/ultravox_model.py:547]: Is this what you meant?
```suggestion
    def init_latency_mask(self, audio_latency_block_size: int, dtype: torch.dtype):
```

**Review by farzadab — COMMENTED** [review]
Seems like `just test` failed since `mypy` is complaining.

I proposed the solution for two of the issues. The last one I couldn't figure out at a glance why the value could be None, but it should be easy to figure out with the code editor's help.

**Comment by saeeddhqan** [comment]
@farzadab solved.
